[meerkatboy] iP

meerkatboy opened the pull request:
# Meerkatbot
> "I am a meerkat"  - A meerkat

Meerkatbot frees your mind of having to remember things you need to do. It's
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,
1. download it from [here](https://nus-cs2103-ay2425s1.github.io/website/schedule/week4/project.html)
2. double-click it
3. add your tasks
4. let meerkat manage your tasks for you 😉 

And it is **FREE!**

Features:
- [x] Managing tasks
- [ ] Managing deadlines
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

meerkatboy committed:
changed file name, changed greeting

meerkatboy committed:
added echo feature, changed goodbye text

meerkatboy committed:
added store to list, displays stored items when list is input

meerkatboy committed:
added task function, ability to mark as completed or incomplete. created new task class

meerkatboy committed:
created 3 new classes todo, event, deadline, edited popup message

meerkatboy committed:
added automated text ui testcase, ran checks

meerkatboy committed:
added try catch blocks to catch some exceptions and wrongful input

meerkatboy committed:
added a delete task function

meerkatboy committed:
refactore code to categorise code into methods, write to list of tasks file whenever changes are made

meerkatboy committed:
ran some direct tests

meerkatboy committed:
Merge branch 'branch-Level-7'

meerkatboy committed:
group methods and fields into separate classes
previously all under Meerkat.java, not grouped into relevant classes
missed the date and time assignment, will do it after this

meerkatboy committed:
group methods and fields into separate classes
previously all under Meerkat.java, not grouped into relevant classes
missed the date and time assignment, will do it after this

meerkatboy committed:
Merge tag 'A-MoreOOP'

meerkatboy committed:
Merge branch 'master' of https://github.com/meerkatboy/ip

meerkatboy committed:
Merge branch 'add-gradle-support'

meerkatboy committed:
add gradle files

limyuquan commented on IP/src/main/java/Parser.java:
```diff
@@ -0,0 +1,153 @@
+
+    private static TaskList taskList;
+    private static Ui ui;
+    private final String STORAGEFILEPATH = "Meerkat.txt";
```
words should be separated by _

meerkatboy replied:
Okay! Have made the edits, thank you!

limyuquan commented on IP/src/main/java/Storage.java:
```diff
@@ -0,0 +1,31 @@
+
+public class Storage {
+
+    private final String STORAGEFILEPATH = "Meerkat.txt";
```
same issue

limyuquan commented on IP/src/main/java/Task.java:
```diff
@@ -0,0 +1,39 @@
+public class Task {
+    private boolean completed;
```
should be named isXXXfor boolean

meerkatboy replied:
Noted! Thanks!

limyuquan commented on IP/src/main/java/TaskList.java:
```diff
@@ -0,0 +1,96 @@
+    private Storage storage;
+    private Ui ui;
+    private static List<Task> listOfTasks = new ArrayList<>();
+    private final String STORAGEFILEPATH = "Meerkat.txt";
```
same issue

limyuquan commented on IP/src/main/java/Meerkat.java:
```diff
@@ -0,0 +1,25 @@
+
+public class Meerkat {
+
+    private static final String STORAGEFILEPATH = "Meerkat.txt";
```
word should be separated by _

Anvita2110 replied:
>[!NOTE]
>Constant names must be all uppercase using underscore to separate words (aka SCREAMING_SNAKE_C

limyuquan reviewed (commented):
other than these small coding standard issues, looks good

meerkatboy committed:
Merge remote-tracking branch 'remotes/origin/add-gradle-support' into HEAD

Anvita2110 commented on IP/src/main/java/Deadline.java:
```diff
@@ -0,0 +1,24 @@
+public class Deadline extends Task {
+    private String duedate;
```
Shouldn't the variable name (duedate) be in camelCase?

meerkatboy replied:
It should, have made the necessary edit, thank you!

Anvita2110 commented on IP/src/main/java/Event.java:
```diff
@@ -0,0 +1,26 @@
+        else {
+            s += "u,";
+        }
+        s += this.getName() + "," + this.start + "," + this.end;
```
Perhaps using String.format( ) would be easier here than concatenating?

Anvita2110 commented on IP/src/main/java/Parser.java:
```diff
@@ -0,0 +1,153 @@
+
+    private static TaskList taskList;
+    private static Ui ui;
+    private final String STORAGEFILEPATH = "Meerkat.txt";
```
>[!NOTE]
>Constant names must be all uppercase using underscore to separate words (aka SCREAMING_SNAKE_C

Anvita2110 commented on IP/src/main/java/Parser.java:
```diff
@@ -0,0 +1,153 @@
+            ui.printNoIdeaMessage();
+        }
+    }
+}
```
Well done on the exception handling part, variable names are very descriptive
### **Good job!**

Anvita2110 commented on IP/src/main/java/Task.java:
```diff
@@ -0,0 +1,39 @@
+public class Task {
+    private boolean completed;
```
>[!NOTE]
> Boolean variables/methods should be named to sound like booleans
> for example:  isFinished, isFound, isOpen, hasData, wasOpen

Anvita2110 commented on IP/src/main/java/TaskList.java:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    public String taskListToString(List<Task> taskList) {
+        StringBuilder s = new StringBuilder();
```
Would suggest to use stringBuilder variable name rather than just 's' to avoid ambiguity with mostly Scanner object.

meerkatboy replied:
Noted, will implement the changes! Thank you :)

Anvita2110 commented on IP/src/main/java/TaskList.java:
```diff
@@ -0,0 +1,96 @@
+            listOfTasks.get(size - 1).markAsIncomplete();
+        }
+    }
+}
```
Neatly written code, exception handling stands out! 😊

Anvita2110 commented on IP/src/main/java/Ui.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    public void printTaskCreationMessage(Task thisTask, TaskList taskList) {
+        System.out.println(lines + "\nGot it. I've added this task:\n  " + thisTask + "\nNow you have " + taskList.getSize() + " tasks in the list\n" + lines);
```
Don't know the exact number of characters but this line could be indented by entering to next line.

Anvita2110 commented on IP/src/main/java/Ui.java:
```diff
@@ -0,0 +1,81 @@
+    public void printErrorWritingFileMessage() {
+        System.out.println(lines + "\nerror in writing to file\n" + lines);
+    }
+
```
Great job in handling cases well!

Anvita2110 reviewed (commented):
## Well done with functionality

* Code is well written
* Great job at exception handling
* Variable names and method names are well descriptive!

Overall, did a good job, with few suggestions and mainly constant name error in a few places so do look out for them as they are used multiple times in different files.

meerkatboy committed:
Add Gradle Support

The previous version of the chatbot did not support building and running using Gradle

Merged the add-gradle-support remote branch and added the relevant files to support Gradle build

meerkatboy committed:
Add JUnit support and DateTime functionality.

Programme before commit did not support running JUnit tests and previous DateTime implementation was deleted on accident.

Current program parses string input into LocalDateTime and supports JUnit tests

meerkatboy committed:
Merge branch 'branch-A-JUnit'

meerkatboy committed:
Package chatbot as a jar and fixed a bug.

Before commit, chatbot did not exist as a jar and bug caused all tasks to be marked as undone.

Tasks are now correctly marked based on save file after editing switch statements, chatbot exists as jar.

meerkatboy committed:
Add JavaDoc functionality.

Code beforehand could have been confusing without JavaDoc annotation.

Now with annotation, other readers will be able to determine code functionality clearly.

meerkatboy committed:
Modify code to adhere to coding standards

Some constants were not in screaming snake case and some unnecessary
variables were declared.

Removed unnecessary variables and modified constants to screaming snake
case.

meerkatboy committed:
Modify UI class to restrict code line length

A line of code was 160 characters long.

Modified to shorten it to adhere to coding standards.

meerkatboy committed:
Lack function to find tasks based on keyword

Longer search time to identify tasks

Providing find functionality increases efficiency

Let's add a function where users can input keyword and a list of
tasks matching the keyword will be displayed

meerkatboy committed:
Add CheckStyle function

Has no CheckStyle functionality so we may stray away from coding
standards

Adhering to coding standards is essential for ensuring code readability

Running checkstyle tests with Gradle ensures checks in code style is
consistently run

meerkatboy committed:
Add GUI to chatbot

Chatbot had been solely text based

Text based chatbot can be less engaging and fun for users

Use of JavaFx to create GUI enhances user experience

meerkatboy committed:
Modify GUI to make it more aesthetic

Was very plan and unengaging

Improved design will better attract target audience usage

meerkatboy committed:
Refactor code to improve code CodeQuality

Method lengths are too long in some cases

Shorten the code through extraction of code blocks into separate methods

Makes code easier to read and debug

meerkatboy committed:
Refactor code to reduce method length

Duplicates in tasklist can confuse the user.

By implementing function to detect duplicates, user can no longer add
a task if an identical task is already present in the taskList.

meerkatboy committed:
Finalise function to detect duplicates

Duplicates can be very confusing for the user when they look back at
the task list, especially when they intend to make any edits.

Finalising the function to detect duplicates to prevent them from being
added in the first place can help to avoid that problem.

meerkatboy committed:
Add information into README regarding features

Chatbot features can be difficult to navigate for first time users.
Having a README can help one explore the features of the chatbot, to
utilise it more effectively.

Detailed descriptions of the features, as well as example usages
are reflected to make the chatbot more easy to understand and use.

meerkatboy committed:
Add to feature list in README

Include other features that were previously not included. This makes
the README more comprehensive.

meerkatboy committed:
Finalise user guide and debug program

User guide is important for users to fully understand how to use the
application. A working program without bugs is essential for good
user experience.

Code refactoring is implemented to improve use of OOP concepts. Bugs
are also rectified to enhance the user experience.

meerkatboy committed:
Update README to include png image

Having a Ui png image can make the guide more easily understandable.

Changing the filepath can ensure the image is shown on the user guide.
